File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/InviteManageViewModel.kt
// P6 — drives the per-app invite manage screen. Mirrors
// FlagshipUI/ViewModels/InviteManageViewModel.swift 1:1.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.AppInviteAccessSummary
import com.flagshipserver.app.api.AppInvitePendingSummary
import com.flagshipserver.app.api.AppInviteRevokeRequest
import com.flagshipserver.app.api.ScreensClient
import com.flagshipserver.app.api.ScreensError
import com.flagshipserver.app.core.InviteLabel
import com.flagshipserver.app.core.InviteLabelBook
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class InviteManageViewModel(
    val serviceId: String,
    private val client: ScreensClient,
    private val labelBook: InviteLabelBook,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {
    data class Snapshot(
        val pending: List<AppInvitePendingSummary>,
        val access: List<AppInviteAccessSummary>,
    )

    private val _state = MutableStateFlow<LoadingState<Snapshot>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<Snapshot>> = _state.asStateFlow()

    private val _revokePending = MutableStateFlow(false)
    val revokePending: StateFlow<Boolean> = _revokePending.asStateFlow()

    private val _lastRevokeOutcome = MutableStateFlow<String?>(null)
    val lastRevokeOutcome: StateFlow<String?> = _lastRevokeOutcome.asStateFlow()

    fun load(): Job = scope.launch {
        _state.value = LoadingState.Loading
        runCatching {
            coroutineScope {
                val pending = async { client.appInviteList(serviceId) }
                val access = async { client.appInviteAccess(serviceId) }
                Pair(pending.await(), access.await())
            }
        }.fold(
            onSuccess = { (pendingResp, accessResp) ->
                _state.value = LoadingState.Loaded(
                    Snapshot(pending = pendingResp.pending, access = accessResp.access),
                )
            },
            onFailure = { t -> _state.value = LoadingState.Failed(failureMessage(t)) },
        )
    }

    /** Resolve the local label for a given opaqueTag. Returns null when
     *  the issuance happened on another device. */
    fun label(opaqueTagHex: String): InviteLabel? =
        labelBook.get(serviceId, opaqueTagHex)

    fun revokeInvite(inviteId: String, opaqueTagHex: String?): Job =
        runRevoke(AppInviteRevokeRequest.invite(serviceId, inviteId), opaqueTagHex)

    fun revokeAccess(irkPubKey: String, opaqueTagHex: String?): Job =
        runRevoke(AppInviteRevokeRequest.access(serviceId, irkPubKey), opaqueTagHex)

    private fun runRevoke(req: AppInviteRevokeRequest, localTag: String?): Job = scope.launch {
        _revokePending.value = true
        try {
            runCatching { client.appInviteRevoke(req) }.fold(
                onSuccess = { resp ->
                    if (!localTag.isNullOrEmpty()) {
                        labelBook.remove(serviceId, localTag)
                    }
                    _lastRevokeOutcome.value =
                        if (resp.alreadyRevoked == true) "already revoked" else "revoked"
                    load().join()
                },
                onFailure = { t -> _lastRevokeOutcome.value = failureMessage(t) },
            )
        } finally {
            _revokePending.value = false
        }
    }

    private fun failureMessage(t: Throwable): String = when (t) {
        is ScreensError.Http -> t.message ?: "HTTP ${t.status}"
        else -> t.message ?: "couldn't load invites"
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/InviteManageViewModel.swift
import Foundation
import Observation
import FlagshipAPI
import FlagshipCore

/// P6 — drives the per-app invite manage screen. Fans out two reads
/// (`appInviteList` + `appInviteAccess`) in parallel and exposes a
/// revoke action that re-fetches both lists on success.
@MainActor
@Observable
public final class InviteManageViewModel {
    public struct Snapshot: Equatable, Sendable {
        public var pending: [AppInvitePendingSummary]
        public var access: [AppInviteAccessSummary]
        public init(pending: [AppInvitePendingSummary], access: [AppInviteAccessSummary]) {
            self.pending = pending
            self.access = access
        }
    }

    public private(set) var state: LoadingState<Snapshot> = .idle
    public private(set) var revokePending: Bool = false
    public private(set) var lastRevokeOutcome: String?

    public let serviceId: String
    private let client: any ScreensClient
    private let labelBook: any InviteLabelBook

    public init(serviceId: String, client: any ScreensClient, labelBook: any InviteLabelBook) {
        self.serviceId = serviceId
        self.client = client
        self.labelBook = labelBook
    }

    public func load() async {
        state = .loading
        do {
            async let pendingTask = client.appInviteList(serviceId: serviceId)
            async let accessTask = client.appInviteAccess(serviceId: serviceId)
            let (pendingResp, accessResp) = try await (pendingTask, accessTask)
            state = .loaded(Snapshot(pending: pendingResp.pending, access: accessResp.access))
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    /// Resolve the local label for a given opaqueTag. Returns nil when
    /// the issuance happened on another device (or after a wipe).
    public func label(for opaqueTagHex: String) -> InviteLabel? {
        labelBook.get(serviceId: serviceId, opaqueTagHex: opaqueTagHex)
    }

    public func revokeInvite(inviteId: String, opaqueTagHex: String?) async {
        await runRevoke(.invite(serviceId: serviceId, inviteId: inviteId), localTag: opaqueTagHex)
    }

    public func revokeAccess(irkPubKey: String, opaqueTagHex: String?) async {
        await runRevoke(.access(serviceId: serviceId, irkPubKey: irkPubKey), localTag: opaqueTagHex)
    }

    private func runRevoke(_ req: AppInviteRevokeRequest, localTag: String?) async {
        revokePending = true
        defer { revokePending = false }
        do {
            let resp = try await client.appInviteRevoke(req)
            if let tag = localTag, !tag.isEmpty {
                labelBook.remove(serviceId: serviceId, opaqueTagHex: tag)
            }
            lastRevokeOutcome = resp.alreadyRevoked == true ? "already revoked" : "revoked"
            await load()
        } catch {
            lastRevokeOutcome = error.localizedDescription
        }
    }
}
